src/AI/Muriel.java

```diff
@@ -405,8 +405,30 @@ public GameAction getNextAction(GameMap gameMap)
           UnitMatchupAndMetaInfo umami = getUnitMatchupInfo(unit, target);
           if( umami.costEffectivenessRatio < COST_EFFECTIVENESS_MIN ) continue;
 
+          // Find locations that would be dangerous for us so we can avoid sauntering into enemy fire.
+          HashSet<XYCoord> noGoZone = new HashSet<XYCoord>();
+          final int MAX_RELEVANT_DISTANCE = 27; // ~3x the move distance of the fastest units. Up for tweaking.
+          for(int j = 0; j < enemyLocations.size(); ++j)
+          {
+            XYCoord threatCoord = enemyLocations.get(j);
+            Unit threat = gameMap.getLocation(threatCoord).getResident();
+            XYCoord unitCoord = new XYCoord(unit.x, unit.y);
+            if( Utils.findManhattanDistance(unitCoord, threatCoord) <= MAX_RELEVANT_DISTANCE )
+            {
+              // If we, in the enemy's place, would attack `unit` with `threat`, then we should not let them attack us.
+              UnitMatchupAndMetaInfo invUmami = getUnitMatchupInfo(threat, unit);
+              if( invUmami.costEffectivenessRatio > COST_EFFECTIVENESS_MIN && threat.getHP() >= 6 )
+              {
+                // Add coordinates that `threat` could target to our "no-go" list.
+                Map<XYCoord, Double> threatMap = AIUtils.findThreatPower(gameMap, threat, unit.model);
+                noGoZone.addAll(threatMap.keySet()); // Ignore the valueMap of the return; we have already decided `threat` is dangerous.
```

Comment:
This does ignore the effect of terrain and the multiple-weapons-with-different-ranges thingie.
That's not really important at the moment, though.

Might be worth breaking `findThreatPower()` into two parts: "where can this threaten" and "how threatening, really?"
Then again, that would be difficult to do right since you'd have to reverse-engineer the conditions under which the unit in question could threaten the tile you'd like to diagnose.
It's interesting looking at my past design decisions and going "this could be better" and then "wait nevermind, past me has a point"

Reply:
> effect of terrain

One day I intend to account for that. I have ideas. Not today.

And yeah, it can be hard to break logic apart without forcing yourself to repeat some of it.